src/circuit.rs

```diff
@@ -120,6 +121,49 @@ pub struct Circuit {
     pub(crate) rcv: Value<ValueCommitTrapdoor>,
 }
 
+impl Circuit {
+    /// Constructs a `Circuit` from the following components:
+    /// - `merkle_path`: a merkle path witness of the `input_note`
+    /// - `input_note`: a note spent in scope of the action
+    /// - `output_note`: a note created in scope of the action
+    /// - `fvk`: full viewing key owning the `input_note`
+    /// - `alpha`: a scalar used for randomization of the action spend validating key
+    /// - `rcv`: trapdoor for the action value commitment
+    ///
+    /// Returns `None` if the `input_note` is not owned by the `fvk`.
+    pub fn from_action_context(
+        merkle_path: tree::MerklePath,
+        input_note: Note,
+        output_note: Note,
```

Comment:
Passing in the notes like this means that we don't have a guarantee that the output note's `rho` is using the input note's nullifier. Add a check that these are the same (which will require deriving the nullifier for the input note).

Reply:
done in 603a5cf798fe82d174c514d16f79ef70cc4a4eb5